drivers/wdm/audio/backpln/portcls/pin_wavert.cpp

```diff
@@ -305,35 +305,46 @@ CPortPinWaveRT::DeviceIoControl(
     IoStack = IoGetCurrentIrpStackLocation(Irp);
 
 
-    if (IoStack->Parameters.DeviceIoControl.IoControlCode == IOCTL_KS_PROPERTY)
+    switch (IoStack->Parameters.DeviceIoControl.IoControlCode)
     {
-       return HandleKsProperty(Irp);
+        case IOCTL_KS_PROPERTY:
+            return HandleKsProperty(Irp);
+            break;
+		
+        case IOCTL_KS_ENABLE_EVENT:
+            /* FIXME UNIMPLEMENTED */
+            UNIMPLEMENTED_ONCE;
+            break;
+
+        case IOCTL_KS_DISABLE_EVENT:
+            /* FIXME UNIMPLEMENTED */
+            UNIMPLEMENTED_ONCE;
+            break;
+
+        case IOCTL_KS_HANDSHAKE:
+            /* FIXME UNIMPLEMENTED */
+            UNIMPLEMENTED_ONCE;
+            break;
+
+        case IOCTL_KS_METHOD:
+            /* FIXME UNIMPLEMENTED */
+            UNIMPLEMENTED_ONCE;
+            break;
+
+        case IOCTL_KS_RESET_STATE:
+            /* FIXME UNIMPLEMENTED */
+            UNIMPLEMENTED_ONCE;
+            break;
+			
+        case IOCTL_KS_WRITE_STREAM:
+        case IOCTL_KS_READ_STREAM:
+            return HandleKsStream(Irp);
+            break;
+			
+        default:
```

Comment:
Only these 8 `IOCTL_KS_*` are defined, so shouldn't we add a `DPRINT1("unreachable\n")` or rework this code further?

Reply:
It's not unreachable, user mode can send arbitrary values. `KsDefaultDeviceIoCompletion` will correctly handle this case by completing the IRP with `STATUS_INVALID_DEVICE_REQUEST`.

Reply:
Thanks for the explanation. `STATUS_INVALID_DEVICE_REQUEST` is what made me wonder.